modules/dnn/src/layers/scatterND_layer.cpp

```diff
@@ -0,0 +1,220 @@
+            CV_CheckEQ(inputs[2][i + m], inputs[0][j + m], "ScatterND: updates.shape[rank(indices)-1 : ] must equal to data[indices.shape[-1] : rank(data)-1].");
+        }
+
+        outputs.assign(1, inputs[0]);
+        return false;
+    }
+
+    virtual void finalize(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr) CV_OVERRIDE
+    {
+        std::vector<Mat> inputs;
+        inputs_arr.getMatVector(inputs);
+
+        int ndims = inputs[0].dims;
+        inp_current_indices.resize(ndims);
+    }
+
+    void forward(InputArrayOfArrays inputs_arr, OutputArrayOfArrays outputs_arr, OutputArrayOfArrays internals_arr) CV_OVERRIDE
+    {
+        CV_TRACE_FUNCTION();
+        CV_TRACE_ARG_VALUE(name, "name", name.c_str());
+
+        std::vector<Mat> inputs, outputs;
+        inputs_arr.getMatVector(inputs);
+        outputs_arr.getMatVector(outputs);
+
+        const Mat& data = inputs[0];
+        const Mat& indices = inputs[1];
+        const Mat& updates = inputs[2];
+        Mat& out = outputs[0];
+
+        typeDispatch(outputs[0].type(), data, indices, updates, out);
+    }
+
+    // NOTE: This impl does not check whether indices have duplicate entries.
+    //       The last duplicate entry will overwrite the previous.
+    template<typename T, typename Functor>
+    void forward_impl(const Functor& rd, const Mat& data, const Mat& indices, const Mat& updates, Mat& out)
+    {
+        data.copyTo(out);
+
+        const int ndims = data.dims;
+        const int* shape = data.size.p;
+        const size_t* step = data.step.p;
+
+        const int ind_ndims = indices.dims;
+        const int* ind_shape = indices.size.p;
+        const T* p_indices = indices.ptr<const T>();
+
+        const int upd_ndims = updates.dims;
+        const int* upd_shape = updates.size.p;
+        const T* p_updates = updates.ptr<const T>();
+
+        T* p_out = out.ptr<T>();
+
+        int k = ind_shape[ind_ndims - 1]; // last dim of indices
+        size_t total = (size_t)(indices.total() / k);
+
+        size_t updates_size = 1;
+        for (int i = ind_ndims - 1; i < upd_ndims; i++)
+            updates_size *= upd_shape[i];
+
+        size_t inp_start_offset = 0;
+        size_t ind_start_offset = 0;
+        size_t upd_start_offset = 0;
+        for (size_t i = 0; i < total; i++, ind_start_offset += k, upd_start_offset += updates_size)
+        {
+            const T* tmp_p_indices = p_indices + ind_start_offset;
+            std::fill(inp_current_indices.begin(), inp_current_indices.end(), 0);
+            for (int j = 0; j < k; j++)
+            {
+                CV_Assert(tmp_p_indices[j] < shape[j] && tmp_p_indices[j] > -shape[j]);
+                inp_current_indices[j] = ((int)tmp_p_indices[j] + shape[j]) % shape[j];
+            }
+
+            inp_start_offset = 0;
+            for (int j = 0; j < ndims; j++)
+                inp_start_offset += inp_current_indices[j] * step[j];
+            inp_start_offset /= sizeof(T);
```

Comment:
I think we can get rid of `inp_current_indices` and `finalize()`:
```cpp
const T* tmp_p_indices = p_indices + ind_start_offset;
inp_start_offset = 0;
for (int j = 0; j < k; j++)
{
    CV_Assert(tmp_p_indices[j] < shape[j] && tmp_p_indices[j] > -shape[j]);
    inp_start_offset += (((int)tmp_p_indices[j] + shape[j]) % shape[j]) * step[j];
}
inp_start_offset /= sizeof(T);
```
It shows another 2x speedup on my machine.

Reply:
It is indeed ~2X speedup getting rid of `inp_current_indices`